designs/webviews/webviews-general.md

```diff
@@ -0,0 +1,95 @@
+
+## Tenets (unless you know better ones)
+
+1. Users should have a delightful experience when interacting with AWS features through our AWS Toolkit.
+2. We should not limit ourselves to IDE-specific conventions if it does not provide the best user experience possible.
+3. We should do our best to not compromise the VS Code editor's or our toolkit's performance; using our toolkit should be an easy choice for an AWS developer to make, rather than a compromise.
+
+## Current State
+
+VS Code offers a very limited set of UI functionality. Specifically, VS Code offers the following constructs for user-interactible workflows:
+
+* [Quick Picks](https://code.visualstudio.com/api/extension-capabilities/common-capabilities#quick-pick), which present a list of options for a user to select. Users can select one or multiple items and can filter the list but cannot add to it.
+* [Input Boxes](https://code.visualstudio.com/api/extension-capabilities/common-capabilities#quick-pick), which allow a user to submit a string.
+* [File Pickers](https://code.visualstudio.com/api/extension-capabilities/common-capabilities#file-picker), which allow a user to choose a file or directory.
+* [Modal/Non-Modal/Status Bar Messages](https://code.visualstudio.com/api/references/vscode-api#window), which can contain arbitrary buttons.
+* [Tree Views](https://code.visualstudio.com/api/references/vscode-api#TreeView), representing hierarchical information.
+
+The workflow-centric constructs (quick picks, input boxes, and file pickers) can only be displayed one at a time, which can lead to confusion in workflows where users are sequentially asked for inputs. For instance, a previous step's selection is invalid could cause a current selection to become invalid, Backtracking through the workflow to correct the mistake (or simply change values) is a pain point. These workflows also remove the user's context, forcing them to backtrack to remember what they entered previously.
+
+Another major reported pain point from working within the VS Code-specific constructs come from some of the confusion with how the inputs look the same but register differently depending on their type; an example can be seen in [the following Github issue](https://github.com/aws/aws-toolkit-vscode/issues/650).
+
+Our toolkit does currently include some webviews based on Vue.js. These webviews for the most part get the job done but have an inconsistent design and have little to no styling. These webviews are also written in Javascript, eschewing any type safety. Furthermore, they use a relatively esoteric Lodash-based templating system to create their HTML. We have limited their use within the toolkit, both due to these facts as well as a desire to stay as close to native VS Code constructs as possible; the only workflows that use these are for remote AWS Lambda invocations and EventBridge Schema searching, which both feature use-cases that are too complex for the workflow constructs.
```

Comment:
This is dropping back into implementation - this doc should be focussing on the required customer experience.

Reply:
This paragraph and the next paragraph could be replaced with something like "Searching EventBridge Schemas is one example of webview-based functionality. Creating a new SAM Application is one exaple of a sequential-input experience."

I don't know what purpose those sentences would serve though. Maybe it would make sense above line 30 (after the VS Code component bullet list)